rizinorg/cutter
# Add reverse debugging

**Detailed description**

Basic reverse debug features, currently WIP since it's broken in all kinds of ways on r2's side for emulation(major issues) and all debug types(small bugs for each plugin).

![reversedebug](https://user-images.githubusercontent.com/5659696/70820436-96492300-1dd0-11ea-9475-4ee97add37cb.PNG)
![reverdebugmenu](https://user-images.githubusercontent.com/5659696/70820435-96492300-1dd0-11ea-9038-b1cb8d6d1c83.PNG)


**Test plan (required)**

* Start and stop a tracing session multiple times.
* See that sessions are stopped when you exit debug.
* Start a trace session, step/continue forwards and then start stepping backwards
* Start a trace session, move forward to a certain point and see that continue backwards returns to the instruction you started tracing at or the nearest breakpoint
* Check that the register values are correctly restored after continue/step back

**Closing issues**

closes #1022
